bdk_rpc_example/src/main.rs

```diff
@@ -0,0 +1,240 @@
+        } => {
+            let (chan, recv) = sync_channel::<RpcData>(CHANNEL_BOUND);
+            let sigterm_flag = start_ctrlc_handler(chan.clone());
+            let local_cps = keychain_tracker.chain().checkpoints().clone();
+
+            // emit blocks thread
+            let join_handle = std::thread::spawn(move || loop {
+                client.emit_blocks(&chan, &local_cps, fallback_height)?;
+                if live && !await_flag(&sigterm_flag, Duration::from_secs(10)) {
+                    continue;
+                }
+                return chan.send(RpcData::Stop(true)).map_err(RpcError::Send);
+            });
+
+            let mut tip = 0;
+
+            for data in recv.iter() {
+                let mut update = ChainGraph::<TxHeight>::default();
+                let is_synced = matches!(&data, &RpcData::Mempool(_));
+
+                let txs = match data {
+                    RpcData::Start {
+                        local_tip,
+                        target_tip,
+                    } => {
+                        tip = target_tip;
+                        println!(
+                            "sync start: current_tip={}, target_tip={}",
+                            local_tip, target_tip
+                        );
+                        continue;
+                    }
+                    RpcData::Stop(finished) => {
+                        println!("terminating... sync_finished={}", finished);
+                        drop(recv);
+                        break;
+                    }
+                    RpcData::Blocks { cp, blocks } => {
+                        let checkpoints =
+                            blocks.iter().map(|(h, b)| (*h, b.block_hash())).chain(cp);
+                        for (height, hash) in checkpoints {
+                            update.insert_checkpoint(BlockId { height, hash })?;
+                        }
+
+                        blocks
+                            .into_iter()
+                            .flat_map(|(height, block)| {
+                                block
+                                    .txdata
+                                    .into_iter()
+                                    .map(move |tx| (TxHeight::Confirmed(height), tx))
+                            })
+                            .collect::<Vec<_>>()
+                    }
+                    RpcData::Mempool(txs) => txs
+                        .into_iter()
+                        .map(|tx| (TxHeight::Unconfirmed, tx))
+                        .collect::<Vec<_>>(),
+                };
+
+                for (height, tx) in txs {
+                    keychain_tracker
+                        .txout_index
+                        .derive_until_unused_gap(stop_gap);
+                    if keychain_tracker.txout_index.is_relevant(&tx) {
+                        update.insert_tx(tx.clone(), Some(height))?;
+                    }
+                    keychain_tracker.txout_index.scan(&tx);
```

Comment:
I think this is unsound -- you have to scan the txs first before deciding if it's relevant. scan the block when it comes in and each tx individually when the mempool comes in. From the docs of `is_relevant`:

```
    /// Whether any of the inputs of this transaction spend a txout tracked or whether any output
    /// matches one of our script pubkeys.
    ///
    /// It is easily possible to misuse this method and get false negatives by calling it before you
    /// have scanned the `TxOut`s the transaction is spending. For example if you want to filter out
    /// all the transactions in a block that are irrelevant you **must first scan all the
    /// transactions in the block** and only then use this method.
```

Reply:
But we are scanning transactions in order so the previous outputs will be scanned before the current transaction right?

Reply:
Oh that's a good point but that only applies to the block txs. In the mempool are you guaranteed some kind of logical traversal order when you get the response. Maybe? But we shouldn't depend on it.